object_store/src/instrumentation.rs

```diff
@@ -0,0 +1,1102 @@
+    // which point the metrics are emitted and this field is set to None,
+    // preventing the drop impl duplicating them.
+    last_yielded_at: Option<Time>,
+    // The error state of the last poll - true if OK, false if an error
+    // occurred.
+    //
+    // This is used to select the correct success/error histogram which records
+    // the operation duration.
+    last_call_ok: bool,
+
+    get_success_duration: U64Histogram,
+    get_error_duration: U64Histogram,
+    get_bytes: U64Counter,
+}
+
+impl<S> StreamMetricRecorder<S>
+where
+    S: Stream,
+{
+    fn new(
+        stream: S,
+        started_at: Time,
+        get_success_duration: U64Histogram,
+        get_error_duration: U64Histogram,
+        get_bytes: U64Counter,
+    ) -> Self {
+        let time_provider = SystemProvider::default();
+        Self {
+            inner: stream,
+
+            // Set the last_yielded_at to now, ensuring the duration of work
+            // already completed acquiring the steam is correctly recorded even
+            // if the stream is never polled / data never read.
+            last_yielded_at: Some(time_provider.now()),
+            // Acquiring the stream was successful, even if the data was never
+            // read.
+            last_call_ok: true,
+
+            started_at,
+            time_provider,
+
+            get_success_duration,
+            get_error_duration,
+            get_bytes,
+        }
+    }
+}
+
+impl<S, P, E> Stream for StreamMetricRecorder<S, P>
+where
+    S: Stream<Item = Result<Bytes, E>>,
+    P: TimeProvider,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        let res = this.inner.poll_next(cx);
+
+        match res {
+            Poll::Ready(Some(Ok(bytes))) => {
+                *this.last_call_ok = true;
+                *this.last_yielded_at.as_mut().unwrap() = this.time_provider.now();
+                this.get_bytes.inc(bytes.len() as _);
+                Poll::Ready(Some(Ok(bytes)))
+            }
+            Poll::Ready(Some(Err(e))) => {
+                *this.last_call_ok = false;
+                *this.last_yielded_at.as_mut().unwrap() = this.time_provider.now();
+                Poll::Ready(Some(Err(e)))
+            }
+            Poll::Ready(None) => {
+                // The stream has terminated - record the wall clock duration
+                // immediately.
+                let hist = match this.last_call_ok {
+                    true => this.get_success_duration,
+                    false => this.get_error_duration,
+                };
+
+                if let Some(d) = this
+                    .last_yielded_at
+                    .expect("no last_yielded_at value for fused stream")
+                    .checked_duration_since(*this.started_at)
+                {
+                    hist.record(d.as_millis() as _)
+                }
```

Comment:
I am not sure it is clearer, but you can avoid the explicitt assignment to `None` if you wanted via `take()`


```suggestion
                if let Some(d) = this
                    .last_yielded_at
                    .take()
                    .expect("no last_yielded_at value for fused stream")
                    .checked_duration_since(*this.started_at)
                {
                    hist.record(d.as_millis() as _)
                }
```